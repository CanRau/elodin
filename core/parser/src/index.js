import Parser from './Parser'
import parse from './parse'
import errorTypes from './errorTypes'

export { Parser, parse, errorTypes }
