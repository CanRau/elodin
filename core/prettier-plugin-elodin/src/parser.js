import { format } from '@elodin/format'

export default function parse(text, _, options) {
  const formatted = format(text)

  const errors = formatted.errors
  let formattedText = formatted.code

  if (errors.length > 0) {
    let formattedText = text
  }

  return {
    ast_type: 'elodin-ast',
    body: formattedText,
    error: errors[0],
    end: text.length,
    source: text,
    start: 0,
  }
}
