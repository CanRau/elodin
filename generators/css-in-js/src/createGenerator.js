import {
  isPseudoClass,
  isPseudoElement,
  isMediaQuery,
  generateCSSMediaQueryFromNode,
  generateCSSClasses,
  getModuleName,
  stringifyCSSRule,
} from '@elodin/utils'
import { isUnitlessProperty, hyphenateProperty } from 'css-in-js-utils'
import color from 'color'

import adapters from './adapters'

export default function createGenerator({
  adapter = 'fela',
  devMode = false,
  rootNode = 'body',
  dynamicImport = false,
} = {}) {
  const usedAdapter = adapters.find(adapt => adapt.name === adapter)
  const config = {
    devMode,
    rootNode,
    dynamicImport,
  }

  return function generate(ast, fileName) {
    const css = generateCSS(ast, config)
    const js = generateJS(ast, config, usedAdapter)
    const root = generateRoot(ast, config)

    return { [fileName + '.js']: root, ...css, ...js }
  }
}

function generateRoot(ast) {
  // TODO: include fragments
  const styles = ast.body.filter(node => node.type === 'Style')

  const imports = styles
    .map(
      module =>
        'import { ' + module.name + " } from './" + module.name + ".elo.js'"
    )
    .join('\n')

  return (
    imports +
    '\n\n' +
    'export {\n  ' +
    styles.map(module => module.name).join(',\n  ') +
    '\n}'
  )
}

function generateCSS(ast, { devMode }) {
  // TODO: include fragments
  const styles = ast.body.filter(node => node.type === 'Style')
  const variants = ast.body.filter(node => node.type === 'Variant')

  return styles.reduce((files, module) => {
    const classes = generateCSSClasses(module.body, variants)

    files[module.name + '.elo.css'] = classes
      .filter(selector => selector.declarations.length > 0)
      .map(selector => {
        const css = stringifyCSSRule(
          selector.declarations,
          getModuleName(module, devMode) + selector.modifier + selector.pseudo,
          selector.media ? '  ' : ''
        )

        if (selector.media) {
          return '@media ' + selector.media + ' {\n' + css + '\n}'
        }

        return css
      })
      .join('\n\n')

    return files
  }, {})
}

function generateJS(ast, { devMode, dynamicImport }, adapter) {
  // TODO: include fragments
  const styles = ast.body.filter(node => node.type === 'Style')
  const variants = ast.body.filter(node => node.type === 'Variant')

  return styles.reduce((files, module) => {
    const style = generateStyle(module.body)
    const classNameMap = generateClassNameMap(module.body, variants)

    files[module.name + '.elo.js'] = adapter.stringify({
      style,
      moduleName: module.name,
      dynamicImport,
      classNameMap,
      className: '_elo_' + module.format + ' ' + getModuleName(module, devMode),
      variants: variants.reduce((flatVariants, variant) => {
        flatVariants[variant.name] = variant.body.map(
          variation => variation.value
        )

        return flatVariants
      }, {}),
    })

    return files
  }, {})
}

function generateClassNameMap(
  nodes,
  variants,
  classes = {},
  variations = {},
  modifier = ''
) {
  const base = nodes.filter(node => node.type === 'Declaration')
  const nesting = nodes.filter(node => node.type !== 'Declaration')

  classes[modifier] = variations

  nesting.forEach(nest => {
    if (nest.property.type === 'Identifier') {
      const variant = variants.find(
        variant => variant.name === nest.property.value
      )

      if (variant) {
        if (nest.value.type === 'Identifier') {
          const variation = variant.body.find(
            variant => variant.value === nest.value.value
          )

          if (variation) {
            generateClassNameMap(
              nest.body,
              variants,
              classes,
              {
                ...variations,
                [variant.name]: variation.value,
              },
              modifier + '__' + variant.name + '-' + variation.value
            )
          }
        }
      } else {
        // TODO: throw
      }
    }
  })

  return classes
}

function generateStyle(nodes) {
  const base = nodes.filter(node => node.type === 'Declaration')
  const nestings = nodes.filter(node => node.type !== 'Declaration')

  const declarations = base
    .filter(decl => decl.dynamic)
    .map(declaration => ({
      property: declaration.property,
      value: declaration.value.value,
    }))

  const nests = nestings
    .map(nest => {
      if (nest.property.type === 'Variable' && nest.property.environment) {
        if (nest.boolean) {
          const pseudoClass = isPseudoClass(nest.property.value)
          const pseudoElement = isPseudoElement(nest.property.value)

          if (pseudoClass || pseudoElement) {
            return {
              property: (pseudoElement ? '::' : ':') + nest.property.value,
              value: generateStyle(nest.body),
            }
          }
        }

        if (isMediaQuery(nest.property.value)) {
          return {
            property: getCSSMediaQueryFromNode(
              nest.value.value,
              nest.property.value,
              nest.operator
            ),
            value: generateStyle(nest.body),
          }
        }
      }
    })
    .filter(nesting => nesting && nesting.value.length > 0)

  return [...declarations, ...nests]
}
