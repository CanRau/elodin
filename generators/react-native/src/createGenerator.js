import { uncapitalizeString, getVariablesFromAST } from '@elodin/utils'
import { hyphenateProperty } from 'css-in-js-utils'
import color from 'color'

const defaultConfig = {
  importFrom: 'react-native',
}

export default function createGenerator(customConfig = {}) {
  const config = {
    ...defaultConfig,
    ...customConfig,
  }
  function generate(ast, fileName = '') {
    const js = generateJS(ast, config)

    return { [fileName + '.js']: js }
  }

  generate.filePattern = ['*.elo.js']
  generate.ignorePattern = ['node_modules']

  return generate
}

function stringifyDeclaration(declaration) {
  const prop = declaration.property + ': '

  if (typeof declaration.value === 'object') {
    return (
      prop + '{' + declaration.value.map(stringifyDeclaration).join(',\n') + '}'
    )
  }

  return prop + declaration.value
}

function generateJS(ast, config) {
  const styles = ast.body.filter(node => node.type === 'Style')

  const modules = styles.reduce((modules, module) => {
    const variables = getVariablesFromAST(module)
    const style = generateStyle(module.body)
    modules[module.name] = { style, variables }
    return modules
  }, {})

  return (
    'import { StyleSheet } from "' +
    config.importFrom +
    '"' +
    '\n\n' +
    'const styles = StyleSheet.create({\n  ' +
    Object.keys(modules)
      .map(
        name =>
          uncapitalizeString(name) +
          ': {\n    ' +
          modules[name].style
            .filter(decl => !decl.dynamic)
            .map(stringifyDeclaration)
            .join(',\n    ') +
          '\n  }'
      )
      .join(',\n  ') +
    '\n})' +
    '\n\n' +
    Object.keys(modules)
      .map(
        name =>
          'export function ' +
          name +
          '(props = {}) {\n  ' +
          (modules[name].style.find(decl => decl.dynamic)
            ? 'const style = {\n    ' +
              modules[name].style
                .filter(decl => decl.dynamic)
                .map(stringifyDeclaration)
                .join(',\n    ') +
              '\n  }\n\n  ' +
              'return StyleSheet.flatten([styles.' +
              uncapitalizeString(name) +
              ', style])'
            : 'return styles.' + uncapitalizeString(name)) +
          '\n}'
      )
      .join('\n\n')
  )
}

function generateStyle(nodes) {
  const base = nodes.filter(node => node.type === 'Declaration')
  getVariablesFromAST

  return base.map(declaration => ({
    property: declaration.property,
    value: generateValue(declaration.value, declaration.property === 'opacity'),
    dynamic: declaration.dynamic,
  }))
}

function wrapInString(value) {
  return '"' + value + '"'
}

function wrapInParens(value) {
  return '(' + value + ')'
}

const inlineFns = {
  add: ' + ',
  sub: ' - ',
  mul: ' * ',
  div: ' / ',
  percentage: true,
}

const stringFns = {
  rgb: true,
  rgba: true,
  hsl: true,
  hsla: true,
}

function generateFunction(node, floatingPercentage = false) {
  if (stringFns[node.callee]) {
    return wrapInString(
      node.callee +
        '(' +
        node.params
          .map(param => {
            if (
              param.type === 'Variable' ||
              (param.type === 'FunctionExpression' && inlineFns[param.callee])
            ) {
              return '" + ' + wrapInParens(generateValue(param, true)) + ' + "'
            }
            return generateValue(param, true)
          })
          .join(', ') +
        ')'
    )
  }

  if (node.callee === 'percentage') {
    if (floatingPercentage) {
      return wrapInParens(
        generateValue(node.params[0], floatingPercentage) + ' / 100'
      )
    }

    return '(' + generateValue(node.params[0], floatingPercentage) + ') + "%"'
  }

  if (node.callee === 'raw') {
    return wrapInString(generateValue(node.params[0], floatingPercentage))
  }

  if (inlineFns[node.callee]) {
    return wrapInParens(
      node.params
        .map(value => generateValue(value, floatingPercentage))
        .join(inlineFns[node.callee])
    )
  }

  // if (math[node.callee]) {
  //   return generateValue({
  //     type: 'Integer',
  //     value: resolveMath(value),
  //   })
  // }
}

function generateValue(node, floatingPercentage = false) {
  if (node.type === 'FunctionExpression') {
    return generateFunction(node, floatingPercentage)
  }

  if (node.type === 'Integer') {
    return (node.negative ? '-' : '') + node.value
  }

  if (node.type === 'Float') {
    return (node.negative ? '-' : '') + node.integer + '.' + node.fractional
  }

  if (node.type === 'Identifier') {
    return wrapInString(hyphenateProperty(node.value))
  }

  if (node.type === 'Variable') {
    return 'props.' + node.value
  }

  return wrapInString(node.value)
}
