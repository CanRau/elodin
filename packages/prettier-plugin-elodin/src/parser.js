import { format } from '@elodin/format'

export default function parse(text) {
  const formattedText = format(text)

  return {
    ast_type: 'elodin-ast',
    body: formattedText,
    end: text.length,
    source: text,
    start: 0,
  }
}
