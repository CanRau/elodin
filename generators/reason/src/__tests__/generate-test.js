import { parse } from '@elodin/parser'

import createGenerator from '../createGenerator'

const file = `
view Button {
  backgroundColor: red
  paddingLeft: 10
  paddingBottom: 10
  paddingTop: $top
}

text ButtonText {
  fontSize: $size
  fontWeight: $weight
  lineHeight: 1.3
  fontFamily: "Arial"
}`

describe('Compiling to ReasonML', () => {
  it('should return a map of files', () => {
    const { ast } = parse(file)

    expect(createGenerator()(ast, 'root.elo')).toMatchSnapshot()
  })
})
