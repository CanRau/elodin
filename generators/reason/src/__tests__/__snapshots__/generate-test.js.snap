// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compiling to ReasonML should return a map of files 1`] = `
Object {
  "RootStyle.re": "open Css;
[%bs.raw {|
  require(\\"./RootStyleButton.elo.css\\")
|}];

[%bs.raw {|
  require(\\"./RootStyleLabel.elo.css\\")
|}];

[%bs.raw{|
  require(\\"./_reset.elo.css\\")
|}];

[@bs.deriving jsConverter]
type type_ =
  | Primary
  | Secondary;

[@bs.deriving jsConverter]
type mode =
  | Dark
  | Light
  | Semi;

let buttonStyle = (~borderWidth, ~paddingLeft, ~paddingRight, ~padLeft) => style([unsafe(\\"border-width\\", borderWidth)]);
let getButtonStyleVariants = (~borderWidth, ~paddingLeft, ~paddingRight, ~padLeft, ~type_, ~mode, ()) => {
    let buttonStylePrimary = style([unsafe(\\"padding-left\\", paddingLeft),
      hover([unsafe(\\"padding-right\\", paddingRight)])]);
    let buttonStylePrimaryDark = style([unsafe(\\"padding-left\\", padLeft)]);

switch (type_, mode) {
    | (Some(Primary), Some(Dark)) => [buttonStylePrimary,buttonStylePrimaryDark]
    | (Some(Primary), Some(Light)) => [buttonStylePrimary]
    | (Some(Primary), Some(Semi)) => [buttonStylePrimary]
    | (Some(Primary), None) => [buttonStylePrimary]
    | (Some(Secondary), Some(Dark)) => []
    | (Some(Secondary), Some(Light)) => []
    | (Some(Secondary), Some(Semi)) => []
    | (Some(Secondary), None) => []
    | (None, Some(Dark)) => []
    | (None, Some(Light)) => []
    | (None, Some(Semi)) => []
    | (None, None) => []
    }
  };

let getButtonVariants = (~type_, ~mode, ()) => {
  switch (type_, mode) {
    | (Some(Primary), Some(Dark)) => \\" _333se4_0-0 _333se4_1-0 _333se4_0-0_1-0\\"
    | (Some(Primary), Some(Light)) => \\" _333se4_0-0 _333se4_1-1 _333se4_0-0_1-1\\"
    | (Some(Primary), Some(Semi)) => \\" _333se4_0-0 _333se4_1-2 _333se4_0-0_1-2\\"
    | (Some(Primary), None) => \\" _333se4_0-0\\"
    | (Some(Secondary), Some(Dark)) => \\" _333se4_0-1 _333se4_1-0 _333se4_0-1_1-0\\"
    | (Some(Secondary), Some(Light)) => \\" _333se4_0-1 _333se4_1-1 _333se4_0-1_1-1\\"
    | (Some(Secondary), Some(Semi)) => \\"\\"
    | (Some(Secondary), None) => \\" _333se4_0-1\\"
    | (None, Some(Dark)) => \\" _333se4_1-0\\"
    | (None, Some(Light)) => \\" _333se4_1-1\\"
    | (None, Some(Semi)) => \\" _333se4_1-2\\"
    | (None, None) => \\"\\"
  }
};

let button = (~borderWidth, ~paddingLeft, ~paddingRight, ~padLeft, ~type_=?, ~mode=?, ()) => {
  \\"_elo_view _333se4\\" ++ getButtonVariants(~type_, ~mode, ()) ++ \\" \\" ++ merge([buttonStyle(~borderWidth, ~paddingLeft, ~paddingRight, ~padLeft), ...getButtonStyleVariants(~borderWidth, ~paddingLeft, ~paddingRight, ~padLeft, ~type_, ~mode, ())]);
}

let labelStyle = (~lineHeight, ~fontSize, ~mediaFontSize, ~mediaHoverFontSize) => style([unsafe(\\"line-height\\", lineHeight),
    hover([unsafe(\\"font-size\\", fontSize)]),
    media(\\"(min-width: 320px) and (max-width: 320px)\\", [unsafe(\\"font-size\\", mediaFontSize),
hover([unsafe(\\"font-size\\", mediaHoverFontSize)])])]);
let getLabelStyleVariants = (~lineHeight, ~fontSize, ~mediaFontSize, ~mediaHoverFontSize, ~mode, ()) => {
    

switch (mode) {
    | (Some(Dark)) => []
    | (Some(Light)) => []
    | (Some(Semi)) => []
    | (None) => []
    }
  };

let getLabelVariants = (~mode, ()) => {
  switch (mode) {
    | (Some(Dark)) => \\"\\"
    | (Some(Light)) => \\" _1imz7a4_0-1\\"
    | (Some(Semi)) => \\"\\"
    | (None) => \\"\\"
  }
};

let label = (~lineHeight, ~fontSize, ~mediaFontSize, ~mediaHoverFontSize, ~mode=?, ()) => {
  \\"_elo_text _1imz7a4\\" ++ getLabelVariants(~mode, ()) ++ \\" \\" ++ merge([labelStyle(~lineHeight, ~fontSize, ~mediaFontSize, ~mediaHoverFontSize), ...getLabelStyleVariants(~lineHeight, ~fontSize, ~mediaFontSize, ~mediaHoverFontSize, ~mode, ())]);
}
",
  "RootStyleButton.elo.css": "._333se4 {
  background-color: red;
  padding-left: 10px;
  border: 0px
}

._333se4_0-0 {
  background-color: red
}

._333se4_0-0:hover {
  padding-left: 10px
}

._333se4_0-0_1-0 {
  background-color: blue
}

._333se4_1-1 {
  background-color: green
}",
  "RootStyleLabel.elo.css": "._1imz7a4:hover {
  color: red
}

._1imz7a4_0-1 {
  font-size: 16px
}

@media (min-width: 320px) and (max-width: 320px) {
  ._1imz7a4 {
    color: green
  }
}

@media (min-width: 320px) and (max-width: 320px) {
  ._1imz7a4:hover {
    color: blue
  }
}",
  "_reset.elo.css": "._elo_view{display:-webkit-box;display:-moz-box;display:-ms-flexbox;display:-webkit-flex;display:flex;overflow:hidden;-webkit-overflow-scrolling:touch;align-self:stretch;flex-direction:column;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-shrink:0;max-width:100%;box-sizing:border-box}._elo_text{display:inline;overflow:visible}",
}
`;

exports[`Compiling to ReasonML should return a map of files in devMode 1`] = `
Object {
  "RootStyle.re": "open Css;
[%bs.raw {|
  require(\\"./RootStyleButton.elo.css\\")
|}];

[%bs.raw {|
  require(\\"./RootStyleLabel.elo.css\\")
|}];

[%bs.raw{|
  require(\\"./_reset.elo.css\\")
|}];

[@bs.deriving jsConverter]
type type_ =
  | Primary
  | Secondary;

[@bs.deriving jsConverter]
type mode =
  | Dark
  | Light
  | Semi;

let buttonStyle = (~borderWidth, ~paddingLeft, ~paddingRight, ~padLeft) => style([unsafe(\\"border-width\\", borderWidth)]);
let getButtonStyleVariants = (~borderWidth, ~paddingLeft, ~paddingRight, ~padLeft, ~type_, ~mode, ()) => {
    let buttonStylePrimary = style([unsafe(\\"padding-left\\", paddingLeft),
      hover([unsafe(\\"padding-right\\", paddingRight)])]);
    let buttonStylePrimaryDark = style([unsafe(\\"padding-left\\", padLeft)]);

switch (type_, mode) {
    | (Some(Primary), Some(Dark)) => [buttonStylePrimary,buttonStylePrimaryDark]
    | (Some(Primary), Some(Light)) => [buttonStylePrimary]
    | (Some(Primary), Some(Semi)) => [buttonStylePrimary]
    | (Some(Primary), None) => [buttonStylePrimary]
    | (Some(Secondary), Some(Dark)) => []
    | (Some(Secondary), Some(Light)) => []
    | (Some(Secondary), Some(Semi)) => []
    | (Some(Secondary), None) => []
    | (None, Some(Dark)) => []
    | (None, Some(Light)) => []
    | (None, Some(Semi)) => []
    | (None, None) => []
    }
  };

let getButtonVariants = (~type_, ~mode, ()) => {
  switch (type_, mode) {
    | (Some(Primary), Some(Dark)) => \\" Button_333se4__Type_-Primary Button_333se4__Mode-Dark Button_333se4__Type_-Primary__Mode-Dark\\"
    | (Some(Primary), Some(Light)) => \\" Button_333se4__Type_-Primary Button_333se4__Mode-Light Button_333se4__Type_-Primary__Mode-Light\\"
    | (Some(Primary), Some(Semi)) => \\" Button_333se4__Type_-Primary Button_333se4__Mode-Semi Button_333se4__Type_-Primary__Mode-Semi\\"
    | (Some(Primary), None) => \\" Button_333se4__Type_-Primary\\"
    | (Some(Secondary), Some(Dark)) => \\" Button_333se4__Type_-Secondary Button_333se4__Mode-Dark Button_333se4__Type_-Secondary__Mode-Dark\\"
    | (Some(Secondary), Some(Light)) => \\" Button_333se4__Type_-Secondary Button_333se4__Mode-Light Button_333se4__Type_-Secondary__Mode-Light\\"
    | (Some(Secondary), Some(Semi)) => \\"\\"
    | (Some(Secondary), None) => \\" Button_333se4__Type_-Secondary\\"
    | (None, Some(Dark)) => \\" Button_333se4__Mode-Dark\\"
    | (None, Some(Light)) => \\" Button_333se4__Mode-Light\\"
    | (None, Some(Semi)) => \\" Button_333se4__Mode-Semi\\"
    | (None, None) => \\"\\"
  }
};

let button = (~borderWidth, ~paddingLeft, ~paddingRight, ~padLeft, ~type_=?, ~mode=?, ()) => {
  \\"_elo_view Button_333se4\\" ++ getButtonVariants(~type_, ~mode, ()) ++ \\" \\" ++ merge([buttonStyle(~borderWidth, ~paddingLeft, ~paddingRight, ~padLeft), ...getButtonStyleVariants(~borderWidth, ~paddingLeft, ~paddingRight, ~padLeft, ~type_, ~mode, ())]);
}

let labelStyle = (~lineHeight, ~fontSize, ~mediaFontSize, ~mediaHoverFontSize) => style([unsafe(\\"line-height\\", lineHeight),
    hover([unsafe(\\"font-size\\", fontSize)]),
    media(\\"(min-width: 320px) and (max-width: 320px)\\", [unsafe(\\"font-size\\", mediaFontSize),
hover([unsafe(\\"font-size\\", mediaHoverFontSize)])])]);
let getLabelStyleVariants = (~lineHeight, ~fontSize, ~mediaFontSize, ~mediaHoverFontSize, ~mode, ()) => {
    

switch (mode) {
    | (Some(Dark)) => []
    | (Some(Light)) => []
    | (Some(Semi)) => []
    | (None) => []
    }
  };

let getLabelVariants = (~mode, ()) => {
  switch (mode) {
    | (Some(Dark)) => \\"\\"
    | (Some(Light)) => \\" Label_1imz7a4__Mode-Light\\"
    | (Some(Semi)) => \\"\\"
    | (None) => \\"\\"
  }
};

let label = (~lineHeight, ~fontSize, ~mediaFontSize, ~mediaHoverFontSize, ~mode=?, ()) => {
  \\"_elo_text Label_1imz7a4\\" ++ getLabelVariants(~mode, ()) ++ \\" \\" ++ merge([labelStyle(~lineHeight, ~fontSize, ~mediaFontSize, ~mediaHoverFontSize), ...getLabelStyleVariants(~lineHeight, ~fontSize, ~mediaFontSize, ~mediaHoverFontSize, ~mode, ())]);
}
",
  "RootStyleButton.elo.css": ".Button_333se4 {
  background-color: red;
  padding-left: 10px;
  border: 0px
}

.Button_333se4__Type_-Primary {
  background-color: red
}

.Button_333se4__Type_-Primary:hover {
  padding-left: 10px
}

.Button_333se4__Type_-Primary__Mode-Dark {
  background-color: blue
}

.Button_333se4__Mode-Light {
  background-color: green
}",
  "RootStyleLabel.elo.css": ".Label_1imz7a4:hover {
  color: red
}

.Label_1imz7a4__Mode-Light {
  font-size: 16px
}

@media (min-width: 320px) and (max-width: 320px) {
  .Label_1imz7a4 {
    color: green
  }
}

@media (min-width: 320px) and (max-width: 320px) {
  .Label_1imz7a4:hover {
    color: blue
  }
}",
  "_reset.elo.css": "._elo_view{display:-webkit-box;display:-moz-box;display:-ms-flexbox;display:-webkit-flex;display:flex;overflow:hidden;-webkit-overflow-scrolling:touch;align-self:stretch;flex-direction:column;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-shrink:0;max-width:100%;box-sizing:border-box}._elo_text{display:inline;overflow:visible}",
}
`;

exports[`Compiling to ReasonML should return a map of files using dynamic imports 1`] = `
Object {
  "RootStyle.re": "open Css;
[%bs.raw{|
  require(\\"./_reset.elo.css\\")
|}];

[@bs.deriving jsConverter]
type type_ =
  | Primary
  | Secondary;

[@bs.deriving jsConverter]
type mode =
  | Dark
  | Light
  | Semi;

let buttonStyle = (~borderWidth, ~paddingLeft, ~paddingRight, ~padLeft) => style([unsafe(\\"border-width\\", borderWidth)]);
let getButtonStyleVariants = (~borderWidth, ~paddingLeft, ~paddingRight, ~padLeft, ~type_, ~mode, ()) => {
    let buttonStylePrimary = style([unsafe(\\"padding-left\\", paddingLeft),
      hover([unsafe(\\"padding-right\\", paddingRight)])]);
    let buttonStylePrimaryDark = style([unsafe(\\"padding-left\\", padLeft)]);

switch (type_, mode) {
    | (Some(Primary), Some(Dark)) => [buttonStylePrimary,buttonStylePrimaryDark]
    | (Some(Primary), Some(Light)) => [buttonStylePrimary]
    | (Some(Primary), Some(Semi)) => [buttonStylePrimary]
    | (Some(Primary), None) => [buttonStylePrimary]
    | (Some(Secondary), Some(Dark)) => []
    | (Some(Secondary), Some(Light)) => []
    | (Some(Secondary), Some(Semi)) => []
    | (Some(Secondary), None) => []
    | (None, Some(Dark)) => []
    | (None, Some(Light)) => []
    | (None, Some(Semi)) => []
    | (None, None) => []
    }
  };

let getButtonVariants = (~type_, ~mode, ()) => {
  switch (type_, mode) {
    | (Some(Primary), Some(Dark)) => \\" _333se4_0-0 _333se4_1-0 _333se4_0-0_1-0\\"
    | (Some(Primary), Some(Light)) => \\" _333se4_0-0 _333se4_1-1 _333se4_0-0_1-1\\"
    | (Some(Primary), Some(Semi)) => \\" _333se4_0-0 _333se4_1-2 _333se4_0-0_1-2\\"
    | (Some(Primary), None) => \\" _333se4_0-0\\"
    | (Some(Secondary), Some(Dark)) => \\" _333se4_0-1 _333se4_1-0 _333se4_0-1_1-0\\"
    | (Some(Secondary), Some(Light)) => \\" _333se4_0-1 _333se4_1-1 _333se4_0-1_1-1\\"
    | (Some(Secondary), Some(Semi)) => \\"\\"
    | (Some(Secondary), None) => \\" _333se4_0-1\\"
    | (None, Some(Dark)) => \\" _333se4_1-0\\"
    | (None, Some(Light)) => \\" _333se4_1-1\\"
    | (None, Some(Semi)) => \\" _333se4_1-2\\"
    | (None, None) => \\"\\"
  }
};

let button = (~borderWidth, ~paddingLeft, ~paddingRight, ~padLeft, ~type_=?, ~mode=?, ()) => {
  [%bs.raw {| import(\\"./RootStyleButton.elo.css\\") |}];
  \\"_elo_view _333se4\\" ++ getButtonVariants(~type_, ~mode, ()) ++ \\" \\" ++ merge([buttonStyle(~borderWidth, ~paddingLeft, ~paddingRight, ~padLeft), ...getButtonStyleVariants(~borderWidth, ~paddingLeft, ~paddingRight, ~padLeft, ~type_, ~mode, ())]);
}

let labelStyle = (~lineHeight, ~fontSize, ~mediaFontSize, ~mediaHoverFontSize) => style([unsafe(\\"line-height\\", lineHeight),
    hover([unsafe(\\"font-size\\", fontSize)]),
    media(\\"(min-width: 320px) and (max-width: 320px)\\", [unsafe(\\"font-size\\", mediaFontSize),
hover([unsafe(\\"font-size\\", mediaHoverFontSize)])])]);
let getLabelStyleVariants = (~lineHeight, ~fontSize, ~mediaFontSize, ~mediaHoverFontSize, ~mode, ()) => {
    

switch (mode) {
    | (Some(Dark)) => []
    | (Some(Light)) => []
    | (Some(Semi)) => []
    | (None) => []
    }
  };

let getLabelVariants = (~mode, ()) => {
  switch (mode) {
    | (Some(Dark)) => \\"\\"
    | (Some(Light)) => \\" _1imz7a4_0-1\\"
    | (Some(Semi)) => \\"\\"
    | (None) => \\"\\"
  }
};

let label = (~lineHeight, ~fontSize, ~mediaFontSize, ~mediaHoverFontSize, ~mode=?, ()) => {
  [%bs.raw {| import(\\"./RootStyleLabel.elo.css\\") |}];
  \\"_elo_text _1imz7a4\\" ++ getLabelVariants(~mode, ()) ++ \\" \\" ++ merge([labelStyle(~lineHeight, ~fontSize, ~mediaFontSize, ~mediaHoverFontSize), ...getLabelStyleVariants(~lineHeight, ~fontSize, ~mediaFontSize, ~mediaHoverFontSize, ~mode, ())]);
}
",
  "RootStyleButton.elo.css": "._333se4 {
  background-color: red;
  padding-left: 10px;
  border: 0px
}

._333se4_0-0 {
  background-color: red
}

._333se4_0-0:hover {
  padding-left: 10px
}

._333se4_0-0_1-0 {
  background-color: blue
}

._333se4_1-1 {
  background-color: green
}",
  "RootStyleLabel.elo.css": "._1imz7a4:hover {
  color: red
}

._1imz7a4_0-1 {
  font-size: 16px
}

@media (min-width: 320px) and (max-width: 320px) {
  ._1imz7a4 {
    color: green
  }
}

@media (min-width: 320px) and (max-width: 320px) {
  ._1imz7a4:hover {
    color: blue
  }
}",
  "_reset.elo.css": "._elo_view{display:-webkit-box;display:-moz-box;display:-ms-flexbox;display:-webkit-flex;display:flex;overflow:hidden;-webkit-overflow-scrolling:touch;align-self:stretch;flex-direction:column;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-shrink:0;max-width:100%;box-sizing:border-box}._elo_text{display:inline;overflow:visible}",
}
`;

exports[`Compiling to ReasonML should return no css files 1`] = `
Object {
  "RootStyle.re": "open Css;


[%bs.raw{|
  require(\\"./_reset.elo.css\\")
|}];

[@bs.deriving jsConverter]
type type_ =
  | Primary
  | Secondary;

[@bs.deriving jsConverter]
type mode =
  | Dark
  | Light
  | Semi;

let buttonStyle = (~borderWidth, ~paddingLeft, ~paddingRight, ~padLeft) => style([unsafe(\\"background-color\\", \\"red\\"),
    unsafe(\\"padding-left\\", \\"10px\\"),
    unsafe(\\"border\\", \\"0px\\"),
    unsafe(\\"border-width\\", borderWidth)]);
let getButtonStyleVariants = (~borderWidth, ~paddingLeft, ~paddingRight, ~padLeft, ~type_, ~mode, ()) => {
    let buttonStylePrimary = style([unsafe(\\"background-color\\", \\"red\\"),
      unsafe(\\"padding-left\\", paddingLeft),
      hover([unsafe(\\"padding-left\\", \\"10px\\"),
unsafe(\\"padding-right\\", paddingRight)])]);
    let buttonStylePrimaryDark = style([unsafe(\\"background-color\\", \\"blue\\"),
      unsafe(\\"padding-left\\", padLeft)]);
    let buttonStyleLight = style([unsafe(\\"background-color\\", \\"green\\")]);

switch (type_, mode) {
    | (Some(Primary), Some(Dark)) => [buttonStylePrimary,buttonStylePrimaryDark]
    | (Some(Primary), Some(Light)) => [buttonStylePrimary,buttonStyleLight]
    | (Some(Primary), Some(Semi)) => [buttonStylePrimary]
    | (Some(Primary), None) => [buttonStylePrimary]
    | (Some(Secondary), Some(Dark)) => []
    | (Some(Secondary), Some(Light)) => [buttonStyleLight]
    | (Some(Secondary), Some(Semi)) => []
    | (Some(Secondary), None) => []
    | (None, Some(Dark)) => []
    | (None, Some(Light)) => [buttonStyleLight]
    | (None, Some(Semi)) => []
    | (None, None) => []
    }
  };

let button = (~borderWidth, ~paddingLeft, ~paddingRight, ~padLeft, ~type_=?, ~mode=?, ()) => {
  \\"_elo_view\\" ++ \\" \\" ++ merge([buttonStyle(~borderWidth, ~paddingLeft, ~paddingRight, ~padLeft), ...getButtonStyleVariants(~borderWidth, ~paddingLeft, ~paddingRight, ~padLeft, ~type_, ~mode, ())]);
}

let labelStyle = (~lineHeight, ~fontSize, ~mediaFontSize, ~mediaHoverFontSize) => style([unsafe(\\"line-height\\", lineHeight),
    hover([unsafe(\\"color\\", \\"red\\"),
unsafe(\\"font-size\\", fontSize)]),
    media(\\"(min-width: 320px) and (max-width: 320px)\\", [unsafe(\\"color\\", \\"green\\"),
unsafe(\\"font-size\\", mediaFontSize),
hover([unsafe(\\"font-size\\", mediaHoverFontSize),
unsafe(\\"color\\", \\"blue\\")])])]);
let getLabelStyleVariants = (~lineHeight, ~fontSize, ~mediaFontSize, ~mediaHoverFontSize, ~mode, ()) => {
    let labelStyleLight = style([unsafe(\\"font-size\\", \\"16px\\")]);

switch (mode) {
    | (Some(Dark)) => []
    | (Some(Light)) => [labelStyleLight]
    | (Some(Semi)) => []
    | (None) => []
    }
  };

let label = (~lineHeight, ~fontSize, ~mediaFontSize, ~mediaHoverFontSize, ~mode=?, ()) => {
  \\"_elo_text\\" ++ \\" \\" ++ merge([labelStyle(~lineHeight, ~fontSize, ~mediaFontSize, ~mediaHoverFontSize), ...getLabelStyleVariants(~lineHeight, ~fontSize, ~mediaFontSize, ~mediaHoverFontSize, ~mode, ())]);
}
",
  "_reset.elo.css": "._elo_view{display:-webkit-box;display:-moz-box;display:-ms-flexbox;display:-webkit-flex;display:flex;overflow:hidden;-webkit-overflow-scrolling:touch;align-self:stretch;flex-direction:column;-webkit-box-orient:vertical;-webkit-box-direction:normal;flex-shrink:0;max-width:100%;box-sizing:border-box}._elo_text{display:inline;overflow:visible}",
}
`;
